### PR TITLE
UTY-709 Add interfaces to be used by MonoBehaviours

### PR DESCRIPTION
**Pull request by firtina-improbable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Introducing the public and internal versions of the Reader interfaces.

Also preparing tagging for component updates to highlight that they are for a particular component. This will aid with type hinting for the rest of MonoBehaviour implementation.

To follow in following PRs:
- events
- commands

#### Tests
Tests to come in a following PR about UTY-709

#### Documentation
Inline documentation for interfaces

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Delegates.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using Improbable.Worker;
+
+namespace Improbable.Gdk.Core.MonoBehaviours
+{
+    // TODO reviewers: shall we put these delegates into their own class/namespace?
```
Comment (MPeti01) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Delegates.cs]: Is there any particular reason they shouldn't be in the same namespace as the other GameObject/MonoBehaviour things just like that? I guess we could have some .Delegates namespace but it looks fine to me as it is too. (Having a class just to keep them in seems weird.)

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Delegates.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    public delegate void AuthorityChangedDelegate(Authority newAuthority);
+
+    public delegate void ComponentUpdateDelegate<in TComponentUpdate>(TComponentUpdate updateData)
```
Comment (MPeti01) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Delegates.cs]: Is there any reason for the contravariant parameter here?

File: workers/unity/Packages/com.improbable.gdk.core/Components/ISpatialComponentUpdate.cs [hunk 3]
```diff
@@ -1,6 +1,13 @@
+using Unity.Entities;
+
 namespace Improbable.Gdk.Core
 {
     public interface ISpatialComponentUpdate
```
Comment (MPeti01) [inline workers/unity/Packages/com.improbable.gdk.core/Components/ISpatialComponentUpdate.cs]: Do we need both the non-generic and the generic versions of the interface?

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/Components/ISpatialComponentUpdate.cs]: The non-generic one is there for convenience where we may care that *it is an update* but not *update of what component*

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Delegates.cs (hunk 2 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Delegates.cs]: Only the recommendation from rider

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Delegates.cs (hunk 1 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Delegates.cs]: Oversight

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/IReaderInternal.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+namespace Improbable.Gdk.Core.MonoBehaviours
+{
+    /// <summary>
+    ///     The interface that is used to signal a reader that a message has been received from the SpatialOS Dispatcher
+    ///     about the component that the reader is connected to.
+    /// </summary>
+    internal interface IReaderInternal
+    {
+        void OnAuthorityChange(Authority authority);
+        void OnComponentUpdate();
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/IReaderInternal.cs]: Heads up: This interface here might be subject to change depending on the discussion in https://github.com/spatialos/UnityGDK/pull/121

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Delegates.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿using Improbable.Worker;
+
+namespace Improbable.Gdk.Core
+{
+    // TODO reviewers: shall we put these delegates into their own class/namespace?
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Delegates.cs]: Yeah I think so.. `GameObjectDelegates` ?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/IReader.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Improbable.Gdk.Core.MonoBehaviours
+{
+    /// <summary>
+    ///     The Reader interface is used by MonoBehaviours to query information about a SpatialOS component and register
+    ///     callbacks for updates, events and commands for component instances.
+    /// </summary>
+    /// <typeparam name="TComponent">The data type for the SpatialOS component.</typeparam>
+    /// <typeparam name="TComponentUpdate">The data type for the SpatialOS component's update.</typeparam>
+    public interface IReader<TComponent, TComponentUpdate>
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/IReader.cs]: Do we need the second generic here?

Comment (MPeti01) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/IReader.cs]: Probably could be removed, and where it's used instead have ISpatialComponentUpdate<TComponent> unless I'm missing something?

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/IReader.cs]: Aye, not necessary, doing the recommended change

**Comment by firtina-improbable** [comment]
For reviewing, example usage:

```cs
// this is a IReader<TestComponentData>
TestComponentData.Reader myReader = reader;

Debug.Log($"Component authority: ${myReader.Authority}");

myReader.AuthorityChanged += authority => { Debug.Log($"Authority changed: ${authority}"); };

myReader.ComponentUpdated += (TestComponentData.Update updateData) =>
{
    if (updateData.Horizontal.HasValue)
    {
        Debug.Log($"New horizontal value: {updateData.Horizontal.Value}");
    }
};
```

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/IReader.cs [hunk 7]
```diff
@@ -7,13 +7,14 @@ namespace Improbable.Gdk.Core.MonoBehaviours
...
     /// </summary>
-    /// <typeparam name="TComponent">The data type for the SpatialOS component.</typeparam>
-    public interface IReader<TComponent>
-        where TComponent : ISpatialComponentData, IComponentData
+    /// <typeparam name="TSpatialComponentData">The data type for the SpatialOS component.</typeparam>
+    public interface IReader<TSpatialComponentData>
+        where TSpatialComponentData : ISpatialComponentData
     {
         Authority Authority { get; }
+        TSpatialComponentData Data { get; }
```
Comment (MPeti01) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/IReader.cs]: Would users of specific subtypes see this property giving a general TSpatialComponentData in the API?

**Pull request merged**